src/npz.rs

```diff
@@ -243,3 +253,545 @@ impl<R: Read + Seek> NpzReader<R> {
+/// file within the archive.
+///
+/// This example uses the [`memmap2`](https://crates.io/crates/memmap2) crate
+/// because that appears to be the best-maintained memory-mapping crate at the
+/// moment, but [`Self::new`] takes a `&mut [u8]` instead of a file so that you
+/// can use the memory-mapping crate you're most comfortable with.
+///
+/// # Example
+///
+/// ```
+/// use std::fs::OpenOptions;
+/// use memmap2::MmapOptions;
+/// use ndarray_npy::{NpzViewMut, ViewNpzError};
+/// use ndarray::Ix1;
+///
+/// // Open `.npz` archive of uncompressed `.npy` files in native endian.
+/// #[cfg(target_endian = "little")]
+/// let mut file = OpenOptions::new().read(true).write(true)
+///     .open("tests/examples_little_endian_64_byte_aligned.npz").unwrap();
+/// #[cfg(target_endian = "big")]
+/// let mut file = OpenOptions::new().read(true).write(true)
+///     .open("tests/examples_big_endian_64_byte_aligned.npz").unwrap();
+/// // Memory-map `.npz` archive of 64-byte aligned `.npy` files.
+/// let mut mmap = unsafe { MmapOptions::new().map_mut(&file).unwrap() };
+/// let mut npz = NpzViewMut::new(&mut mmap)?;
+/// // List uncompressed files only.
+/// for npy in npz.names() {
+///     println!("{}", npy);
+/// }
+/// // Get mutable `.npy` views of both arrays at the same time.
+/// let mut x_npy_view_mut = npz.by_name("i64.npy")?;
+/// let mut y_npy_view_mut = npz.by_name("f64.npy")?;
+/// // Optionally verify CRC-32 checksums.
+/// x_npy_view_mut.verify()?;
+/// y_npy_view_mut.verify()?;
+/// // Get and print mutable `ArrayViewMut`s.
+/// let x_array_view_mut = x_npy_view_mut.view_mut::<i64, Ix1>()?;
+/// let y_array_view_mut = y_npy_view_mut.view_mut::<f64, Ix1>()?;
+/// println!("{}", x_array_view_mut);
+/// println!("{}", y_array_view_mut);
+/// // Update CRC-32 checksums after changes. Automatically updated on `drop()`.
+/// x_npy_view_mut.update();
+/// y_npy_view_mut.update();
+/// # Ok::<(), ViewNpzError>(())
+/// ```
+#[derive(Debug)]
+pub struct NpzViewMut<'a> {
+    files: HashMap<usize, NpyViewMut<'a>>,
+    names: HashMap<String, usize>,
+}
+
+impl<'a> NpzViewMut<'a> {
+    /// Creates a new mutable view of a memory-mapped `.npz` file.
+    pub fn new(mut bytes: &'a mut [u8]) -> Result<Self, ViewNpzError> {
+        let mut zip = ZipArchive::new(Cursor::new(&bytes))?;
+        let mut names = HashMap::new();
+        let mut ranges = HashMap::new();
+        let mut splits = BTreeMap::new();
+        let mut index = 0;
+        for zip_index in 0..zip.len() {
+            let file = zip.by_index(zip_index)?;
+            // Skip directories and compressed files.
+            if file.is_dir() || file.compression() != CompressionMethod::Stored {
+                continue;
+            }
+            // Store file index by file names.
+            let name = file.name().to_string();
+            names.insert(name, index);
+            // Get local CRC-32 range.
```

Comment:
Yep. We also have to test for an optional 4-byte data descriptor signature directly in front of the 4-byte CRC-32 checksum. I wonder if a simple comparison against the signature is enough, that is only the case if the signature itself is not a valid CRC-32 checksum.

Reply:
It seems that the signature itself is a valid CRC-32 checksum. So we have to assume that `0x08074B50` could be a checksum as well. But in that case, we can compare against the central CRC-32 value. If it matches, `0x08074B50` is a checksum, otherwise it's the signature or a malformed archive and we assume that the archive is not malformed.

Reply:
It's a bit more tricky. The specification is broken from the point at which it adopted the signature. To summarize the problematic path:

It is unlikely but quite possible to have a data descriptor. If there is a data descriptor, it is likely to find a signature as it is common practice and recommended. It is **very** unlikely that the signature matches the central CRC-32. But if it does, we would have to compare the next fields (file size) but that route has no ending and gets more and more involved (ZIP64, encryption). Instead I suggest to consider the archive malformed with an appropriate error message to suggest regenerating the archive with a local header instead of an ambiguous data descriptor. This error will probably never happen but it makes the code correct and compact.

Reply:
Done.